common/functions/src/scalars/strings/bin.rs

```diff
@@ -14,103 +14,158 @@
 
 use std::fmt;
 
-use common_datavalues::prelude::*;
-use common_datavalues::DataTypeAndNullable;
+use common_datavalues2::prelude::*;
 use common_exception::ErrorCode;
 use common_exception::Result;
 
-use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::cast_column_field;
 use crate::scalars::function_factory::FunctionFeatures;
-use crate::scalars::Function;
+use crate::scalars::Function2;
+use crate::scalars::Function2Description;
 
 #[derive(Clone)]
 pub struct BinFunction {
     _display_name: String,
 }
 
 impl BinFunction {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function2>> {
         Ok(Box::new(BinFunction {
             _display_name: display_name.to_string(),
         }))
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create))
+    pub fn desc() -> Function2Description {
+        Function2Description::creator(Box::new(Self::try_create))
             .features(FunctionFeatures::default().deterministic().num_arguments(1))
     }
 }
 
-impl Function for BinFunction {
+impl Function2 for BinFunction {
     fn name(&self) -> &str {
         "bin"
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        if !args[0].is_numeric() && !args[0].is_null() {
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        let dt = if args[0].is_nullable() {
+            remove_nullable(args[0]).data_type_id()
+        } else {
+            args[0].data_type_id()
+        };
+
+        if !dt.is_numeric() && !dt.is_string() && !dt.is_null() {
```

Comment:
We can strictly accept numeric data types. 

 String is not allowed, null is covered in `pass_through_nullable `

Reply:
Old wrong tests can be removed.

Reply:
get it.